xapi/futures/managing-cluster-stacks/managing-cluster-stacks.md

```diff
@@ -0,0 +1,168 @@
+not be called across-hosts and therefore do not have to be backward compatible.
+These are "cluster plugin APIs".
+
+We will define the following APIs:
+
+- `Plugin:Membership.create`: add a host to a cluster. On exit the local host cluster software
+  will know about the new host but it may need to be restarted before the
+  change takes effect
+  - in:`cluster:string`: the uuid of the cluster
+  - in:`hostname:string`: the hostname of the management domain
+  - in:`uuid:string`: a UUID identifying the host
+  - in:`id:int`: the lowest available unique integer identifying the host
+      where an integer will never be re-used unless it is guaranteed that
+      all nodes have forgotten any previous state associated with it
+  - in:`configuration:Map(String,String)`: the configuration field of the cluster row.
+  - in:`address:string list`: a list of addresses through which the host
+      can be contacted
+  - out: unit
+- `Plugin:Membership.destroy`: removes a named host from the cluster. On exit the local
+  host software will know about the change but it may need to be restarted
+  before it can take effect  
+  - in:`uuid:string`: the UUID of the host to remove
+  `Plugin:Cluster.query`: queries the state of the cluster
+  - out:`maintenance_required:bool`: true if there is some outstanding configuration
+    change which cannot take effect until the cluster is restarted.
+  - out:`hosts`: a list of all known hosts together with a state including:
+    whether they are known to be alive or dead; or whether they are currently
+    "excluded" because the cluster software needs to be restarted
+- `Plugin:Cluster.start`: turn on the cluster software and let the local host join.
+  This call must be idempotent.
+- `Plugin:Cluster.stop`: turn off the cluster software
+
+Xapi will be modified to:
+
+- add table `Cluster` which will have columns
+  - `name: string`: this is the name of the Cluster plugin (TODO: use same
+    terminology as SM?)
+  - 'configuration: Map(String,String)`: this will contain any cluster-global
+    information, overrides for default values etc.
+  - `enabled: Bool`: this is true when the cluster "should" be running. It
+    may require maintenance to synchronise changes across the hosts.
+  - `maintenance_required: Bool`: this is true when the cluster needs to
+    be placed into maintenance mode to resync its configuration
+- add method `XenAPI:Cluster.enable` which sets `enabled=true` and waits for all
+  hosts to report `Membership.enabled=true`.
+- add method `XenAPI:Cluster.disable` which sets `enabled=false` and waits for all
+  hosts to report `Membership.enabled=false`.
+- add table `Membership` which will have columns
+  - `id: int`: automatically generated lowest available unique integer
+    starting from 0
+  - `cluster: Ref(Cluster)`: the type of cluster. This will never be NULL.
+  - `host: Ref(host)`: the host which is a member of the cluster. This may
```

Comment:
What's the purpose of keeping records of hosts which have left the cluster?   We don't have the hostname or uuid to identify them (because the host ref can be null) so I'm guessing that we can do something with the id.